package/libpiksi/libpiksi/src/networking.c

```
@@ -0,0 +1,248 @@
+    } break;
+    case AF_INET6:
+    {
+      const size_t IP6_SIZE = sizeof(cur_iface->ipv6_address);
+      memcpy(cur_iface->ipv6_address,
+          &((struct sockaddr_in6*)ifa->ifa_addr)->sin6_addr.s6_addr, IP6_SIZE);
+      u8* mask_bytes = (u8*)&((struct sockaddr_in6*)ifa->ifa_netmask)->sin6_addr.s6_addr;
+      cur_iface->ipv6_mask_size = count_set_bits(mask_bytes, IP6_SIZE);
+    } break;
+    case AF_PACKET:
+    {
+      if (ifa->ifa_data != NULL) {
+        struct rtnl_link_stats *stats = ifa->ifa_data;
+        cur_iface->rx_bytes = stats->rx_bytes;
+        cur_iface->tx_bytes = stats->tx_bytes;
+      }
+    } break;
+    default:
+    {
+      // nothing to do here
+    } break;
+  }
+
+  return 0;
+}
+
+void query_network_state(msg_network_state_resp_t *interfaces, u8 interfaces_n, u8 *returned_interfaces)
+{
+  struct fill_network_state_s fill_data = {
+    .interfaces = interfaces,
+    .interfaces_size = interfaces_n,
+    .interfaces_count = returned_interfaces
+  };
+  *returned_interfaces = 0;
+  map_network_interfaces(fill_network_state_struct, (void *)&fill_data);
+}
+
+struct fill_usage_struct_s {
+  network_usage_t *usage_entries;
+  u8 usage_entries_size;
+  u8 *usage_entries_count;
+  u64 duration;
+};
+
+static s32 fill_network_usage_struct(struct ifaddrs *ifa, void *userdata)
+{
+  struct fill_usage_struct_s *fill_usage = (struct fill_usage_struct_s *)userdata;
+  network_usage_t *usage_entries = fill_usage->usage_entries;
+  const u8 usage_entries_size = fill_usage->usage_entries_size;
+  u8 *usage_entries_count = fill_usage->usage_entries_count;
+  u64 duration = fill_usage->duration;
+
+  if (usage_entries == NULL
+      || usage_entries_count == NULL
+      || *usage_entries_count >= usage_entries_size) {
+    return -1;
+  }
+
+  // guard conditions to skip
+  if (ifa->ifa_addr == NULL
+      || ifa->ifa_name == NULL
+      || strlen(ifa->ifa_name) >= sizeof(usage_entries[0].interface_name)) {
```

When would this happen?  Is this a programming error?

---

Assuming you mean the length check, it's valid since we have a limited buffer for the destination.

---

I mean, if we expect this to happen, we should issue a warning, if we don't expect it to happen we should assert/exit if it happens.

---

Since this subroutine is mapped over all the interfaces, returning 0 in this context means `continue`, and returning -1 breaks the mapping loop like a `break`. Are you suggesting that instead of skipping an interface with a name that is too long, that we should stop further interface iteration?

---

Maybe I'm not understanding how the code is used, but I don't want anything to fail silently, this will probably never happen, but let's make a habit to drop a quick log message if it does.